commandapi-core/src/main/java/dev/jorel/commandapi/network/CommandAPIMessenger.java

```diff
@@ -0,0 +1,119 @@
+package dev.jorel.commandapi.network;
+
+import com.google.common.io.ByteArrayDataInput;
+import com.google.common.io.ByteArrayDataOutput;
+import com.google.common.io.ByteStreams;
+
+import java.util.Arrays;
+
+/**
+ * Handles sending and receiving {@link CommandAPIPacket}s between instances of the CommandAPI plugin on different servers.
+ * This class should be extended on each platform to hook into its specific plugin-messaging scheme.
+ *
+ * @param <InputChannel> The type of objects that might send a packet to this plugin
+ * @param <OutputChannel> The type of objects this plugin might send messages to
+ */
+public abstract class CommandAPIMessenger<InputChannel, OutputChannel> {
+	private final CommandAPIPacketHandler<InputChannel> packetHandler;
+
+	/**
+	 * Creates a new {@link CommandAPIMessenger}.
+	 *
+	 * @param packetHandler The {@link CommandAPIPacketHandler} that should handle incoming packets.
+	 */
+	protected CommandAPIMessenger(CommandAPIPacketHandler<InputChannel> packetHandler) {
+		this.packetHandler = packetHandler;
+	}
+
+	/**
+	 * Disables this messenger, unregistering any events and such that let it send/receive packets before the plugin disables.
+	 */
+	public abstract void close();
+
+	/**
+	 * Handles a message sent to this plugin. The given byte array will be decoded into a {@link CommandAPIPacket}
+	 * according to {@link CommandAPIProtocol} and handled appropriately. Nothing happens if the given byte array is empty.
+	 *
+	 * @param sender The source of this message.
+	 * @param input  The byte array that makes up this message.
+	 * @throws IllegalStateException if the packet is malformed and cannot be read correctly.
+	 */
+	protected void messageReceived(InputChannel sender, byte[] input) {
+		// Inspired by net.minecraft.PacketDecoder
+
+		// Empty packet. Kinda weird, but nothing to do.
+		if (input.length == 0) return;
+
+		ByteArrayDataInput buffer = ByteStreams.newDataInput(input);
+
+		int id;
+		CommandAPIPacket packet;
+		try {
+			// Read the id
+			id = buffer.readByte();
+			// Use the id to find and use the correct deserialize method
+			packet = CommandAPIProtocol.createPacket(id, buffer);
+		} catch (IllegalStateException e) {
+			throw new IllegalStateException("Exception while reading packet", e);
+		}
+		if (packet == null) throw new IllegalStateException("Unknown packet id: " + id);
+
+		boolean extraBytes;
+		try {
+			buffer.readByte();
+			extraBytes = true;
+		} catch (IllegalStateException ignored) {
+			// Buffer is out of bytes, as expected
+			// There isn't a method to check this for some reason?
```

Comment:
The `extraBytes` boolean is actually used to 'reverse' the exception.

I want to throw an exception when there are bytes left in the buffer because extra data indicates something went wrong when reading the packet. However, there doesn't seem to be a method that can be used to tell when the buffer is empty. Instead, if reading a byte does not throw an exception, that means there was extra data. When an exception is thrown, the buffer is out of bytes as expected, so we catch that exception and don't throw one ourselves.

This [issue](https://github.com/google/guava/issues/937) indicates a method to check how many bytes are left will not be added. That issue also calls `ByteArrayDataInput` and `ByteArrayDataOutput` a 'failed experiment' 7 years ago, so I'm not sure these classes are _supposed_ to be used anyway. I used them here because that is what the tutorials used. However, I was considering creating a `FriendlyByteBuffer` class similar to what Minecraft uses, and that could easily have a method to check how many bytes are left.